### PR TITLE
Refactoring of heading component

### PR DESCRIPTION
**Pull request by rich-iannone**
This builds upon the earlier refactoring effort, which made available several new utility functions for composing strings. Additional utility functions are included here to help with gathering footnote glyphs and custom styles for cells.

Templating with `glue` is used here to compose the HTML and LaTeX fragments (along with `paste_between()`). Some improvements to the CSS classes for the title and subtitle were also made.

### CODE REVIEW

**Comment by jcheng5** [comment]
I'm going to need you to walk me through this one

File: R/utils_render_html.R [hunk 1]
```diff
@@ -220,169 +266,157 @@ create_table_start_h <- function(groups_rows_df) {
...
-
-      footnote_title_glyphs <-
-        footnote_glyph_to_latex(footnote_title_glyphs$fs_id_coalesced)
-    }
+    footnote_title_glyphs <-
+      switch(output,
+             html = footnote_glyph_to_html(footnote_title_glyphs$fs_id_c),
+             latex = footnote_glyph_to_latex(footnote_title_glyphs$fs_id_c),
+             rtf = footnote_glyph_to_rtf(footnote_title_glyphs$fs_id_c),
+             stop("The context (`", output, "`) is invalid"))
```
Comment (schloerke) [inline R/utils_render_html.R]: have this return the function.  use the resulting fn here and down below.

File: R/cells_stub.R [hunk 2]
```diff
@@ -7,11 +7,6 @@ resolve_location.cells_stub <- function(loc, data_attr) {
   resolved <- resolve_cells_stub(data = stub_df, object = loc)
 
   loc$rows <- resolved$rows
-  # loc$rows <-
-  #   resolve_vars(
-  #     var_expr = loc[["rows"]],
-  #     var_names = stub_df$rowname,
-  #     data_df = stub_df)
```
Comment (schloerke) [inline R/cells_stub.R]: delete

File: R/render_as_html.R [hunk 3]
```diff
@@ -20,83 +18,87 @@ render_as_html <- function(data) {
...
+          html_style = vapply(
+            styles_appended, function(x) styles_to_html(x), character(1))
+        )
+    }
+
+    # # Extraction of body content as a vector
+    # body_content <- as.vector(t(output_df))
+    #
+    # # Split `body_content` by slices of rows
+    # row_splits_body <- split_body_content(body_content, n_cols)
```
Comment (schloerke) [inline R/render_as_html.R]: delete

File: R/summary_rows.R [hunk 4]
```diff
@@ -240,6 +240,22 @@ add_summary_location_row <- function(loc,
            call. = FALSE)
     }
 
+    if (df_type == "footnotes_df") {
+
+      attr(data, df_type) <-
+        add_location_row(
```
Comment (schloerke) [inline R/summary_rows.R]: `add_location_row_footnotes`

File: R/summary_rows.R [hunk 5]
```diff
@@ -240,6 +240,22 @@ add_summary_location_row <- function(loc,
...
+          grpname = group,
+          colname = columns,
+          rownum = rows,
+          text = style
+        )
+
+    } else {
+
     attr(data, df_type) <-
       add_location_row(
```
Comment (schloerke) [inline R/summary_rows.R]: `add_location_row_styles`

File: R/summary_rows.R [hunk 6]
```diff
@@ -299,17 +316,34 @@ add_grand_summary_location_row <- function(loc,
...
-      locnum = 6,
-      grpname = NA_character_,
-      colname = columns,
-      rownum = rows,
-      text = text
-    )
+  if (df_type == "footnotes_df") {
+
+    attr(data, df_type) <-
+      add_location_row(
```
Comment (schloerke) [inline R/summary_rows.R]: `add_location_row_footnotes`

File: R/summary_rows.R [hunk 7]
```diff
@@ -299,17 +316,34 @@ add_grand_summary_location_row <- function(loc,
...
+        grpname = grand_summary_col,
+        colname = columns,
+        rownum = rows,
+        text = style
+      )
+
+  } else {
+
+    attr(data, df_type) <-
+      add_location_row(
```
Comment (schloerke) [inline R/summary_rows.R]: `add_location_row_styles`

File: R/utils_render_html.R [hunk 8]
```diff
@@ -79,310 +23,211 @@ split_body_content <- function(body_content,
...
-      dplyr::mutate(row = as.integer(round((rownum - floor(rownum)) * 100, 0))) %>%
-      dplyr::select(colname, row, styles_appended)
+cell_style_to_html <- function(style) {
+  UseMethod("cell_style_to_html")
+}
 
+cell_style_to_html.default <- function(style) {
 
-    for (i in seq(nrow(styles_summary))) {
+  utils::str(style)
```
Comment (schloerke) [inline R/utils_render_html.R]: delete

Comment (schloerke) [reply R/utils_render_html.R]: nvm this is good

File: R/utils_render_html.R (hunk 1 above)
Comment (schloerke) [reply R/utils_render_html.R]:
```
footnote_title_glyphs_fn <-
  switch(output,
    html = footnote_glyph_to_html,
    latex = footnote_glyph_to_latex,
    rtf = footnote_glyph_to_rtf,
    stop("The context (`", output, "`) is invalid"))
footnote_title_glyphs <- footnote_title_glyphs_fn(footnote_title_glyphs$fs_id_c)
```

File: R/utils_render_html.R [hunk 9]
```diff
@@ -79,310 +23,211 @@ split_body_content <- function(body_content,
...
-      paste0(
-        "<thead>\n<tr>\n",
-        "<th ",
-        "colspan='", n_cols, "' ",
-        "class='gt_heading gt_title gt_font_normal gt_center' ",
-        create_style_attrs(title_style_attrs),
-        ">",
-        heading$title, footnote_title_glyphs,
-        "</th>\n</tr>\n"
+    #title_classes <- "gt_heading gt_title gt_font_normal gt_center"
```
Comment (schloerke) [inline R/utils_render_html.R]: delete

File: R/utils_render_html.R [hunk 10]
```diff
@@ -79,310 +23,211 @@ split_body_content <- function(body_content,
...
+    title_classes <- c("gt_heading", "gt_title", "gt_font_normal", "gt_center")
+
+    #subtitle_classes <- "gt_heading gt_subtitle gt_font_normal gt_center"
```
Comment (schloerke) [inline R/utils_render_html.R]: delete

File: R/utils_general_str_formatting.R [hunk 11]
```diff
@@ -318,3 +318,46 @@ is_adjacent_separate <- function(group_1,
 
   return(TRUE)
 }
+
+str_catalog <- function(item_vector,
```
Comment (schloerke) [inline R/utils_general_str_formatting.R]: where is this function used?

Comment (rich-iannone) [reply R/utils_general_str_formatting.R]: It's used in the `validate_style_in()` function (and I'm thinking of using it elsewhere).

File: R/utils_render_html.R (hunk 10 above)
Comment (rich-iannone) [reply R/utils_render_html.R]: Done.

File: R/utils_render_html.R (hunk 9 above)
Comment (rich-iannone) [reply R/utils_render_html.R]: Done.

File: R/summary_rows.R (hunk 7 above)
Comment (rich-iannone) [reply R/summary_rows.R]: Done.

File: R/summary_rows.R (hunk 6 above)
Comment (rich-iannone) [reply R/summary_rows.R]: Done.

File: R/summary_rows.R (hunk 5 above)
Comment (rich-iannone) [reply R/summary_rows.R]: Done.

File: R/summary_rows.R (hunk 4 above)
Comment (rich-iannone) [reply R/summary_rows.R]: Done.

File: R/render_as_html.R (hunk 3 above)
Comment (rich-iannone) [reply R/render_as_html.R]: Done.

File: R/cells_stub.R (hunk 2 above)
Comment (rich-iannone) [reply R/cells_stub.R]: Done.

File: R/summary_rows.R [hunk 12]
```diff
@@ -243,30 +243,28 @@ add_summary_location_row <- function(loc,
     if (df_type == "footnotes_df") {
 
       attr(data, df_type) <-
```
Comment (schloerke) [inline R/summary_rows.R]: Later, `add_location_row_*` methods might want to return data. (Move the `attr(data, df_type)` inside the function)

Comment (rich-iannone) [reply R/summary_rows.R]: This would be a nice change. Assigning to `attr(data, ...)` sometimes is inconsistent.

Comment (rich-iannone) [reply R/summary_rows.R]: Just made this an issue: https://github.com/rstudio/gt/issues/315

**Pull request merged**